tutorials/how_to/elmo.md

```diff
@@ -0,0 +1,161 @@
+
+Using pre-trained ELMo representations
+--------------------------------------
+
+Pre-trained contextual representations from large scale bidirectional
+language models provide large improvements for nearly all supervised
+NLP tasks.
+
+This document describes how to add ELMo representations to your model using `allennlp`.
+We also have a tensorflow implementation [here](https://github.com/allenai/bilm-tf).
+
+Reference: ["Deep contextualized word representations"](https://openreview.net/forum?id=S1p31z-Ab)
+
+
+## Installing
+
+After installing `allennlp`, download the pre-trained options and weight files.
+
+* [options file](https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_options.json)
+* [weight file](https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_weights.hdf5)
+
+
+## Using ELMo with existing `allennlp` models
+
+In the simplest case, adding ELMo to an existing model is a simple
+configuration change.  We provide a `TokenEmbedder` that accepts
+character ids as input, runs the deep biLM and computes the ELMo representations
+via a learned weighted combination.
+Note that this simple case only includes one layer of ELMo representation
+in the final model.
+In some case (e.g. SQuAD and SNLI) we found that including multiple layers improved performance.  Multiple layers require code changes (see below).
+
+We will use existing SRL model [configuration file](../../training_config/semantic_role_labeler.json) as an example to illustrate the changes.  Without ELMo, it uses 100 dimensional pre-trained GloVe vectors.
+
+To add ELMo, there are three relevant changes.  First, modify the `text_field_embedder` section as follows:
+
+```json
+   "text_field_embedder": {
+     "tokens": {
+       "type": "embedding",
+       "embedding_dim": 100,
+       "pretrained_file": "https://s3-us-west-2.amazonaws.com/allennlp/datasets/glove/glove.6B.100d.txt.gz",
+       "trainable": true
+     },
+     "elmo":{
+       "type": "elmo_token_embedder",
+       "options_file": "/path/to/elmo_2x4096_512_2048cnn_2xhighway_options.json",
```

Comment:
Ideally you could just have an S3 path here as you do for the `pretrained_file`.  Then you can remove your Installing step--making this whole thing a bit easier.  Otherwise everyone has to deal with getting a path mismatch (or at least I do).